### PR TITLE
Support JsonQualifiers with elements.

### PR DESCRIPTION
**Pull request by NightlyNexus**
I need to clean up the functions. They are all nested and confusing. This looks like it works, though.

### CODE REVIEW

**Comment by NightlyNexus** [comment]
Fixes #60.

File: tests/src/main/kotlin/se/ansman/kotshi/ContainsComplexlyQualifiedString.kt [hunk 1]
```diff
@@ -0,0 +1,44 @@
+package se.ansman.kotshi
+
+import com.squareup.moshi.JsonQualifier
+import kotlin.reflect.KClass
+
+// TODO: Elements.getElementValuesWithDefaults(AnnotationMirror) does not find elements with Java keyword names
+// TODO: like `class`, boolean, switch, enum!
```
Comment (NightlyNexus) [inline tests/src/main/kotlin/se/ansman/kotshi/ContainsComplexlyQualifiedString.kt]: Why is this?
When I had `@JsonQualifier annotation class WithBooleanElement(val boolean: Boolean)`, my call to `Elements.getElementValuesWithDefaults(AnnotationMirror)` here https://github.com/ansman/kotshi/pull/62/files#diff-ef2fb2b136b87eda31462cf64df5c75cR217 returned an empty map!

**Comment by ansman** [comment]
I'm a bit curious as to why supporting annotations with arguments isn't straight forward, could you elaborate a bit on that?

**Comment by NightlyNexus** [comment]
Since we have to create our on proxy instance, I had to map the elements' method names to their values and then hand off that map to the created proxy instance.

sorry, i'll get back to this PR tomorrow.

**Comment by NightlyNexus** [comment]
Cleaned up this code for review.

>I'm a bit curious as to why supporting annotations with arguments isn't straight forward, could you elaborate a bit on that?

What do you mean exactly? It's difficult because we have to look at the values in the declared annotation and return them from the proxy instance when the corresponding method is called.
Let me know what I can elaborate on.

**Comment by ansman** [comment]
Could we support this by using reflection to access the annotations on the field rather than create our own instances? To avoid the complex implementation.

File: api/src/main/kotlin/se/ansman/kotshi/KotshiUtils.kt [hunk 2]
```diff
@@ -23,28 +23,34 @@ object KotshiUtils {
             }.append(propertyName)
 
     @JvmStatic
-    fun <T : Annotation> createJsonQualifierImplementation(annotationType: Class<T>): T {
+    fun <T : Annotation> createJsonQualifierImplementation(
+            annotationType: Class<T>, elementNameToArg: Map<String, Any>
```
Comment (ansman) [inline api/src/main/kotlin/se/ansman/kotshi/KotshiUtils.kt]: Put on separate lines

File: api/src/main/kotlin/se/ansman/kotshi/KotshiUtils.kt [hunk 3]
```diff
@@ -23,28 +23,34 @@ object KotshiUtils {
...
                 "hashCode" -> 0
                 "toString" -> "@${annotationType.name}()"
-                else -> method.invoke(proxy, *args)
+                else -> elementNameToArg[method.name] ?: method.invoke(proxy, args)
             }
         } as T
     }
 
+    @JvmStatic
+    fun putInMap(map: MutableMap<String, Any>, key: String, value: Any): MutableMap<String, Any> = map.apply {
```
Comment (ansman) [inline api/src/main/kotlin/se/ansman/kotshi/KotshiUtils.kt]: Make an extension instead? Perhaps with the name `putWithoutConflict` or something similar.

File: compiler/src/main/kotlin/se/ansman/kotshi/AdaptersProcessingStep.kt [hunk 4]
```diff
@@ -201,14 +207,25 @@ class AdaptersProcessingStep(
...
-                                Collections::class.java, LinkedHashSet::class.java, Arrays::class.java)
+                                    Collections::class.java, LinkedHashSet::class.java, Arrays::class.java)
                             .apply {
                                 jsonQualifiers.forEachIndexed { index, qualifier ->
                                     if (index > 0) add(", ")
-                                    add("\$T.createJsonQualifierImplementation(\$T.class)", KotshiUtils::class.java, qualifier)
+                                    val elementValuesWithDefaults = elements.getElementValuesWithDefaults(qualifier)
+                                    if (elementValuesWithDefaults.isEmpty()) {
+                                        add("\$T.createJsonQualifierImplementation(\$T.class, \$T.emptyMap())",
+                                                KotshiUtils::class.java, qualifier.annotationType.asElement(),
```
Comment (ansman) [inline compiler/src/main/kotlin/se/ansman/kotshi/AdaptersProcessingStep.kt]: Put on separate lines.

File: compiler/src/main/kotlin/se/ansman/kotshi/AdaptersProcessingStep.kt [hunk 5]
```diff
@@ -201,14 +207,25 @@ class AdaptersProcessingStep(
...
+                                                Collections::class.java)
+                                    } else {
+                                        add("\$T.createJsonQualifierImplementation(\$T.class, \$T.unmodifiableMap(",
+                                                KotshiUtils::class.java, qualifier.annotationType.asElement(),
```
Comment (ansman) [inline compiler/src/main/kotlin/se/ansman/kotshi/AdaptersProcessingStep.kt]: Put on separate lines.

**Comment by NightlyNexus** [comment]
<details>
<summary>KotshiContainsComplexlyQualifiedStringJsonAdapter</summary>

```java
package se.ansman.kotshi;

import com.squareup.moshi.JsonAdapter;
import com.squareup.moshi.JsonReader;
import com.squareup.moshi.JsonWriter;
import com.squareup.moshi.Moshi;
import java.io.IOException;
import java.lang.Class;
import java.lang.NullPointerException;
import java.lang.Override;
import java.lang.String;
import java.lang.StringBuilder;
import java.util.Arrays;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;

public final class KotshiContainsComplexlyQualifiedStringJsonAdapter extends NamedJsonAdapter<ContainsComplexlyQualifiedString> {
  private static final JsonReader.Options OPTIONS = JsonReader.Options.of(
      "string");

  private final JsonAdapter<String> adapter0;

  public KotshiContainsComplexlyQualifiedStringJsonAdapter(Moshi moshi) {
    super("KotshiJsonAdapter(ContainsComplexlyQualifiedString)");
    adapter0 = moshi.adapter(String.class, Collections.unmodifiableSet(new LinkedHashSet<>(Arrays.asList(KotshiUtils.createJsonQualifierImplementation(WithDefaultStringElement.class, Collections.unmodifiableMap(KotshiUtils.putWithoutConflict(new LinkedHashMap<>(), "string", "default"))), KotshiUtils.createJsonQualifierImplementation(WithArrayElements.class, Collections.unmodifiableMap(KotshiUtils.putWithoutConflict(KotshiUtils.putWithoutConflict(KotshiUtils.putWithoutConflict(new LinkedHashMap<>(), "classArray", new Class<?>[] {}), "byteArray", new byte[] {(byte)0x05}), "stringArray", new String[] {"one", "", "three"}))), KotshiUtils.createJsonQualifierImplementation(WithEnumElement.class, Collections.unmodifiableMap(KotshiUtils.putWithoutConflict(new LinkedHashMap<>(), "someEnum", se.ansman.kotshi.SomeEnum.VALUE5))), KotshiUtils.createJsonQualifierImplementation(WithClassElement.class, Collections.unmodifiableMap(KotshiUtils.putWithoutConflict(new LinkedHashMap<>(), "cls", se.ansman.kotshi.ContainsComplexlyQualifiedString.class))), KotshiUtils.createJsonQualifierImplementation(WithBooleanElement.class, Collections.unmodifiableMap(KotshiUtils.putWithoutConflict(new LinkedHashMap<>(), "bool", true))), KotshiUtils.createJsonQualifierImplementation(WithNumberElement.class, Collections.unmodifiableMap(KotshiUtils.putWithoutConflict(new LinkedHashMap<>(), "number", 4))), KotshiUtils.createJsonQualifierImplementation(WithStringElement.class, Collections.unmodifiableMap(KotshiUtils.putWithoutConflict(new LinkedHashMap<>(), "string", "\\$Hello, ")))))));
  }

  @Override
  public void toJson(JsonWriter writer, ContainsComplexlyQualifiedString value) throws IOException {
    if (value == null) {
      writer.nullValue();
      return;
    }
    writer.beginObject();

    writer.name("string");
    adapter0.toJson(writer, value.getString());

    writer.endObject();
  }

  @Override
  public ContainsComplexlyQualifiedString fromJson(JsonReader reader) throws IOException {
    if (reader.peek() == JsonReader.Token.NULL) {
      return reader.nextNull();
    }

    reader.beginObject();

    String string = null;
    while (reader.hasNext()) {
      switch (reader.selectName(OPTIONS)) {
        case 0: {
          string = adapter0.fromJson(reader);
          continue;
        }
        case -1: {
          reader.nextName();
          reader.skipValue();
          continue;
        }
      }
    }

    reader.endObject();
    StringBuilder stringBuilder = null;
    if (string == null) {
      stringBuilder = KotshiUtils.appendNullableError(stringBuilder, "string");
    }
    if (stringBuilder != null) {
      throw new NullPointerException(stringBuilder.toString());
    }

    return new ContainsComplexlyQualifiedString(
        string);
  }
}
```
</details>

>Could we support this by using reflection to access the annotations on the field rather than create our own instances?

Maybe! Pass in the Field instances to the constructor for fields that have JsonQualifiers and get the qualifiers off them for the adapters created in the constructor? This might be complicated, too. Have any ideas?

**Comment by ansman** [comment]
I mean, wouldn't it just be this: https://github.com/square/moshi/blob/master/moshi/src/main/java/com/squareup/moshi/internal/Util.java#L41

**Comment by NightlyNexus** [comment]
Right, I'm wondering how it'd be best to get the annotation at runtime.
Passing in all the property instances to the JsonAdapter constructor? (messy)

**Comment by ansman** [comment]
I guess the adapter's constructor would have to find the class's constructor, find the right parameter and then the annotations.

**Comment by JakeWharton** [comment]
Isn't the goal of this to avoid reflection?

Why not generate a class which implements the annotation interface, has the
correct equals/hashCode/toString semantics, and new's up instances of them.
Just like AutoAnnotation.

**Comment by ansman** [comment]
Fair enough, I actually didn't realize you would create annotations in runtime without reflection. That would probably be the best, and most efficient solution.

Would that be more efficient for annotations without elements too?

**Comment by JakeWharton** [comment]
More memory efficient but also more classes and methods. So it's a trade
off.

**Comment by ansman** [comment]
No rush, the question about which approach to use still remains.

I think I'm probably leaning towards generating implementations that implement the annotation because it's probably easier to understand and I'm guessing that most projects won't use a lot of qualifiers with methods.

**Comment by NightlyNexus** [comment]
agree, that's what i want to try.

**Comment by ansman** [comment]
I'll take a look at this for 2.0

**Comment by ansman** [comment]
Turns out this is tricker when generating Kotlin. I'll defer this to the next version.

**Comment by ansman** [comment]
Closing this in favor for #128

**Pull request closed without merging**